Add tests and proofs to runtime

jklmnn opened the pull request:
I refactored the source structure and added tests to the runtime libraries.

Then runtime is now divided into `src/` which contains the runtime files (as defined in the current gnat) and `src/lib` which contains any libraries, implementations, specs etc. used to implement runtime functionality and platform interfaces.
The separation was done to create a separate project in `src/lib` which eases proving and testing the implementations. The current runtime library contains:

 - Platform: defines the [platform API](https://github.com/Componolit/ada-runtime/wiki/Platform-interface) and generates the required undefined symbols for the surrounding platform. There are currenlty neither proofs nor tests for the body of this package as I didn't see any reasonable.
 - Ada_Exceptions: defines Exception types and provides them as Ada spec and C header. Both files are autogenerated from `doc/exceptions.md` with `tools/generate_exceptions.py`
 - Ss_Utils: Implementation for Secondary Stack routines. Handles stack allocation and marking. It is proven for the absence of runtime errors. Yet exceptions still might occur in the case that the platform does not provide sufficient resources (e.g. not allocating any memory and returning a null pointer). The tests test the correct behaviour, also for the case of exceptions.
 - String_Utils: Provides a C-Style `strlen` and an Ada to C string conversion. Proven to the point that it will not crash if null pointers are passed and that null pointers never occur internally. Also loops will always increase or exit so no infinite loops exist (in case of endless increasing the loop returns once its index type overflows). The tests check the correct behaviour of length, maximum length and default strings. It is not safe against missing null bytes in C strings (and therefore page faults) so we have to trust the caller.

I also made the `test` build target depend on the runtime. Furthermore I added Aunit to the toolchains used for CI. As the docker container did not provide Aunit I forked it to https://github.com/jklmnn/docker-gnat / https://hub.docker.com/r/jklmnn/gnat/.

treiher commented on src/lib/string_utils.adb:
```diff
@@ -0,0 +1,119 @@
+pragma Ada_2012;
```
Why is this pragma necessary?

treiher commented on src/lib/string_utils.adb:
```diff
@@ -0,0 +1,119 @@
+pragma Ada_2012;
+
+package body String_Utils
+with SPARK_Mode
```
Please fix the indentation.

treiher commented on src/lib/string_utils.adb:
```diff
@@ -0,0 +1,119 @@
+   function Length (
+                    C_Str : System.Address;
+                    Max_Length : Natural := Natural'Last
+                   ) return Integer
```
Could please adapt the style, so that it matches the first or the second example in the [Ada Style Guide](https://en.wikibooks.org/wiki/Ada_Style_Guide/Source_Code_Presentation#More_on_Alignment)? That are the styles which are usually used in the runtime.

treiher commented on src/lib/string_utils.ads:
```diff
@@ -0,0 +1,56 @@
+       Post => Length'Result <= Max_Length,
+       Contract_Cases =>
+         (C_Str = System.Null_Address  => Length'Result <= 0,
+          C_Str /= System.Null_Address => Length'Result >= 0);
```
Wouldn't it make sense to let `Length` return `-1` in case of a null pointer? So it would be possible to distinguish between a null pointer and an empty string. Although it is not strictly necessary in this case, it could be helpful in future applications of this function.

jklmnn replied:
I'm not sure if a distinction is necessary here. In both cases you should't work on the string. I'd rather let the nullpointer check to the user of this function and would not encourage to use this function to check for null pointers explicitly.

senier replied:
Is there a reason not to use a precondition that ensures no null pointer is passed to the function?

jklmnn replied:
Currently I also use it to save a check in the string conversion function (and I need it to initialize the array length so it needs to be able to handle null pointers). When length returns something greater than null I can prove that the address provided to length was not null. So I can check only on the length and do not need to check for a null pointer.

senier replied:
I don't think I understand. Can you point me to the code you're referring to?

jklmnn replied:
First of all the `Length` function is meant to be used in code that we cannot prove or that gets unchecked values from code we have no control over, so it should be able to handle a passed null pointer. So a precondition requiring a valid pointer will fail in this case.

Secondly, if the `Length` function returns any value greater than 0 the pointer passed cannot be null (how else could it have calculated a length then?). On the other side this means that if there has been a null pointer the return value must be zero (but not the other way round, it can be zero even if the pointer is valid).  This is what the contract cases express.

The important part comes at https://github.com/jklmnn/ada-runtime/blob/issue_9_pr/src/lib/string_utils.adb#L38. To get the required length of `Str` I need the length of the string behind the pointer first. To do this at the function declaration part (without a separate declare block) `Length` needs to be able to handle null pointers (as `Convert_To_Ada` is also meant to get untrusted input.
The only case that requires us to process anything is when `L` is greater than zero. The `Get_Char` and `Incr` both have a percondition that they do not accept null pointers. The easy way would be a further check at the beginning of the function (e.g. `if Cursor /= Null_Pointer and then L > 0 then`) but due to the contract cases we already have proven that `Cursor /= Null_Pointer` is true by the fact that `L >0` is also true.

So this whole thing is basically a short cut that reuses the information generated by `Length` instead of rechecking it. I hope this was more understandable.

treiher commented on tests/unit/test.adb:
```diff
@@ -0,0 +1,22 @@
+
+with Rts_Suite;
+
+
```
Please remove duplicate blank lines.

jklmnn commented:
I fixed the coding style with https://github.com/Componolit/ada-runtime/pull/10/commits/0453297b0c4341194570cb97d4b34b02eae30ce8. I'm not sure yet about the return value of `Length`.

jklmnn commented:
As @senier already pointed out we should not have non-proving parts in the SPARK part of the runtime. Yet there are still some exceptions that cause failed checks in ss-utils.adb:
 - [x] [Constraint_Error, L36](https://github.com/jklmnn/ada-runtime/blob/650becf8e443878c20adc85464ae5ea70ebae486/src/lib/ss_utils.adb#L36)
 - [x] [Constraint_Error, L72](https://github.com/jklmnn/ada-runtime/blob/650becf8e443878c20adc85464ae5ea70ebae486/src/lib/ss_utils.adb#L72)
 - [ ] [Storage_Error, L108](https://github.com/jklmnn/ada-runtime/blob/650becf8e443878c20adc85464ae5ea70ebae486/src/lib/ss_utils.adb#L108)
 - [ ] [Program_Error, L138](https://github.com/jklmnn/ada-runtime/blob/650becf8e443878c20adc85464ae5ea70ebae486/src/lib/ss_utils.adb#L138)

Without explaining them in detail the general problem is that we cannot prove the absence of some errors. The last Program_Error could be proven by setting a precondition to `Ss_Release` and assuming that the compiler generated code will never fail to comply the precondition. Yet the other three errors are caused by the constrained number of threads. So we would have to prove that there never happen to be more than 128 (or whatever we choose) threads but as long as we do not dynamically allocate the registry (or make it huge by having it all possible 2**64 - 2 values which is not feasible) I see no way to prove that. A further problem is that there is no way to tell the code calling these function that something failed or that some resources are constrained. So instead of giving away an invalid mark or that of another thread (what would cause undebuggable errors) I decided to fail with an exception at this point.

Generally I'm also not super convinced that making the code prove at any costs is the right way. The reason we have exceptions there is that a situation appeared that we cannot handle at all (what should we do if e.g. there's no memory for the secondary stack?). And "proving" the absence of errors that should be prevented by preconditions that are never checked by the calling code is to my opinion more security by obscurity. What we do is hiding errors that can occur by making an assumption that we know is not always true. This can give a sense of false security and lead to the (wrong) conclusion that this code cannot fail. If we really need to let the prover say that there are no failing checks I would probably let the exceptions as they are and add a pragma (afaik there was one to make the proof work that you need to give an explanation why this proves) that clearly points to our assumption and the problem that can cause the regarding exception.

senier commented:
My point was not to prove the code at all cost (or even add invalid assumption that do not get established).

The issue is that programs with exceptions for which you cannot prove they do *not* happen are not valid SPARK programs. Hence you program is no valid SPARK.

Whenever you need to leave the SPARK boundary, the correct way to do this is a subprogram with `SPARK_Mode => off` for its body. And I guess that's what we should be doing in the case of exceptions. I just don't know whether we can somehow tell the SPARK tools that this is a non-returning function.

senier commented:
[Here](http://docs.adacore.com/live/wave/spark2014/html/spark2014_ug/en/source/language_restrictions.html#raising-exceptions-and-other-error-signaling-mechanisms) is an explanation how to better cope with exceptional conditions in SPARK. I guess what we should be doing (instead of hiding the exception in a non-SPARK subprogram) is adding a precondition to the program that no more than 128 registry entries are requested. Then, we add a check for that number and raise an Exception if it's greater than 128. From the SPARK point of view this can never happen as per the precondition. If it does happen nonetheless, we get an exception as desired.

jklmnn commented:
Okay, this is the mechanism we already use, e.g. at the stack allocation which proves by a postcondition that it never returns a null pointer and raises an exception if this happens.

The current unproven exceptions have the case that they compare the passed arguments against values that are extracted by thread from the registry. I'd prefer to do this together in person as it exceeds my current proving skills.

senier commented:
OK, let's do that together in Monday.

jklmnn commented:
Also you mentioned the following prove errors to me last time:
```
string_utils.adb:22:28: medium: assertion might fail, cannot prove Ptr = Old_Ptr (e.g. when Old_Ptr = 6 and Ptr = 5)
string_utils.adb:42:31: info: function contract not available for proof ("Length" might not return)
string_utils.adb:51:28: medium: assertion might fail, cannot prove Cursor = Old_Cursor (e.g. when Cursor = 3 and Old_Cursor = 4)
```
With gnatprove 18.2 I can prevent them by using codepeer (they only appear without codepeer).
I think someone mentioned that we should include the proofs into the CI. I see two problems there:
 - we only have the GPL version of gnatprove, so no codepeer
 - neither Debian nor the Docker container contain gnatprove

So even ef we had gnatprove available we would need to write all our code to be provable with the GPL version. Also afaik there is still no switch that allows us to set another return code on failed checks.

Nevertheless I created https://github.com/jklmnn/ada-runtime/commit/a718dc8dafc958506b6916ef0e467ba04d8c25f8 to see how such a setup would look like.

jklmnn committed:
prove Constraint_Error ss_utils.adb:41

jklmnn committed:
prove Constraint_Error ss_utils.adb:82

jklmnn committed:
fix mark pre and post conditions

jklmnn committed:
prove get and set mark

jklmnn committed:
mark type invariant

jklmnn commented:
[Constraint_Error, L36](https://github.com/jklmnn/ada-runtime/blob/650becf8e443878c20adc85464ae5ea70ebae486/src/lib/ss_utils.adb#L36) and [Constraint_Error, L72](https://github.com/jklmnn/ada-runtime/blob/650becf8e443878c20adc85464ae5ea70ebae486/src/lib/ss_utils.adb#L72) are fixed with 11a2d805bc862f3b955c1dd2627829f22968196a.

senier commented:
@jklmnn  Please rebase your proof changes onto current master and open another PR.

jklmnn commented:
Is 37a342a intended? As it is a merge commit I cannot rebase this branch to the current master. With this state we could reopen the PR since it does not make any difference and it can be merged directly.

jklmnn commented:
I reopen this due to unresolved questions.

senier commented:
This was my attempt to merge my changes with your commits, but the tests fail after merge. I'd suggest you apply your changes manually (or by cherry-picking) on top of current master.

jklmnn commented:
Okay, I will rebase my changes on master. Should I force push them onto this branch so we don't have to open a new PR? Also the tests fail due to the style checks that fail if the code is compiled as runtime. I intended to fix that after we have proven the last two exceptions.

senier commented:
If force-pushing works as expected, that would be preferable.

Regarding the missing proofs and style checks: I'd prefer if we had a master state without the proof but with succeeding CI. I'd suggest you fix the style issue based on the current state and open an issue for the remaining proofs. Then we can do the proof independently.

jklmnn commented:
I think there is a misunderstanding here. Master is already succeeding on all checks. Only this PR is not (but will be when it is merged to master). This will not change when I force push on this branch.

senier commented:
I thought the errors were introduced by the proofs you added recently. If not, everything should be fine.

jklmnn commented:
They are but they're not on master (this is what I thought was important to you) yet. So this is currently a WIP PR.
I force pushed the new branch. I noticed that you already introduced the early commits to master so only the last five commits (that introduce the build errors) are currently in this pull request.

jklmnn committed:
fix style, remote if expr
ref #10

jklmnn commented:
I tried to fix the failing checks which seemed easy at first, but after all style fixes I was presented with
```
ss_utils.ads:138:27: warning: if expression not allowed in compiler unit
ss_utils.ads:155:07: warning: if expression not allowed in compiler unit
```
These two expressions are the type invariant on the registry entry and the validation check for the registry entry. The validation check could be implemented in a body, but then its properties needed to be denoted at the declaration as a post condition (which requires a if expr...) or a contract case. Since it refers to the elements of the registry entry it needed to be public. This also required the removal of the type invariants of both the registry entry and the mark.
A result that compiles is commit https://github.com/Componolit/ada-runtime/pull/10/commits/0b32bb235609877892cbc452709bc127837520af.
Yet this commit causes a segfault on GNAT GPL 2018 (it calls `Set_Jmpbuf_Address_Soft` and `Get_Jmpbuf_Address_Soft` until it segfaults).

jklmnn commented:
I rebased all changes that worked and fixed their style into https://github.com/jklmnn/ada-runtime/commits/issue_10. Unfortunately these changes only work with GNAT 2018 but not with GNAT 6.3.
The commit introducing the segmentation fault is cd33cdec78a35717ed9866bc212b9aaf7cfb53d7.

jklmnn commented:
Threading support is no longer a requirement for the ss_utils so the proofs for this package are deprecated and will be done after the threading support has been removed.
